rust/src/buffer.rs

```diff
@@ -18,10 +18,74 @@
 use bytes::Bytes;
 use libc;
 use std::mem;
+use std::ptr;
 use std::slice;
 
 use super::memory::*;
 
+/// Buffer builder with zero-copy build method
+pub struct Builder<T> {
+    data: *mut T,
+    len: usize,
+    capacity: usize,
+}
+
+impl<T> Builder<T> {
+    /// Creates a builder with a default capacity
+    pub fn new() -> Self {
+        Builder::with_capacity(64)
+    }
+
+    /// Creates a builder with a fixed capacity
+    pub fn with_capacity(capacity: usize) -> Self {
+        let sz = mem::size_of::<T>();
+        let buffer = allocate_aligned((capacity * sz) as i64).unwrap();
+        Builder {
+            len: 0,
+            capacity,
+            data: unsafe { mem::transmute::<*const u8, *mut T>(buffer) },
+        }
+    }
+
+    /// Push a value into the builder, growing the internal buffer as needed
+    pub fn push(&mut self, v: T) {
+        assert!(!self.data.is_null());
+        if self.len == self.capacity {
+            let sz = mem::size_of::<T>();
+            let new_capacity = self.capacity * 2;
+            unsafe {
+                let old_buffer = self.data;
+                let new_buffer = allocate_aligned((new_capacity * sz) as i64).unwrap();
+                libc::memcpy(
+                    mem::transmute::<*const u8, *mut libc::c_void>(new_buffer),
+                    mem::transmute::<*const T, *const libc::c_void>(old_buffer),
+                    self.len * sz,
+                );
+                self.capacity = new_capacity;
+                self.data = mem::transmute::<*const u8, *mut T>(new_buffer);
+                mem::drop(old_buffer);
+            }
+        }
+        assert!(self.len < self.capacity);
+        unsafe {
+            *self.data.offset(self.len as isize) = v;
+        }
+        self.len += 1;
+    }
+
+    /// Build a Buffer from the existing memory
+    pub fn build(&mut self) -> Buffer<T> {
```

Comment:
Is there a reason that this is called `build`? In C++ we call this `Finish` and it would be nice to have them consistent. If `build` is a typical name used in Rust, it's probably better to keep it as is.